compiler/packages/babel-plugin-react-compiler/src/Optimization/InlineJsxTransform.ts

```diff
@@ -0,0 +1,403 @@
+                keyProperty,
+                propsProperty,
+              ],
+              loc: instr.value.loc,
+            },
+            loc: instr.loc,
+          };
+          nextInstructions.push(reactElementInstruction);
+
+          break;
+        }
+        case 'JsxFragment': {
+          nextInstructions ??= block.instructions.slice(0, i);
+          const {refProperty, keyProperty, propsProperty} =
+            createPropsProperties(
+              fn,
+              instr,
+              nextInstructions,
+              [],
+              instr.value.children,
+            );
+          const reactElementInstruction: Instruction = {
+            id: makeInstructionId(0),
+            lvalue: {...instr.lvalue, effect: Effect.Store},
+            value: {
+              kind: 'ObjectExpression',
+              properties: [
+                createSymbolProperty(
+                  fn,
+                  instr,
+                  nextInstructions,
+                  '$$typeof',
+                  // TODO: Add this to config so we can switch between
+                  // react.element / react.transitional.element
+                  'react.transitional.element',
+                ),
+                createSymbolProperty(
+                  fn,
+                  instr,
+                  nextInstructions,
+                  'type',
+                  'react.fragment',
+                ),
+                refProperty,
+                keyProperty,
+                propsProperty,
+              ],
+              loc: instr.value.loc,
+            },
+            loc: instr.loc,
+          };
+          nextInstructions.push(reactElementInstruction);
+          break;
+        }
+        default: {
+          if (nextInstructions !== null) {
+            nextInstructions.push(instr);
+          }
+        }
+      }
+    }
+    if (nextInstructions !== null) {
+      block.instructions = nextInstructions;
+    }
+  }
+  /**
+   * Fixup the HIR to restore RPO, ensure correct predecessors, and
+   * renumber instructions. Note that the renumbering instructions
+   * invalidates scope and identifier ranges, so we fix them in the
+   * next step.
+   */
+  reversePostorderBlocks(fn.body);
+  markPredecessors(fn.body);
+  markInstructionIds(fn.body);
+
+  /**
+   * Fix scope and identifier ranges to account for renumbered instructions
+   */
+  for (const [, block] of fn.body.blocks) {
+    const terminal = block.terminal;
+    if (terminal.kind === 'scope' || terminal.kind === 'pruned-scope') {
+      /*
+       * Scope ranges should always align to start at the 'scope' terminal
+       * and end at the first instruction of the fallthrough block
+       */
+      const fallthroughBlock = fn.body.blocks.get(terminal.fallthrough)!;
+      const firstId =
+        fallthroughBlock.instructions[0]?.id ?? fallthroughBlock.terminal.id;
+      terminal.scope.range.start = terminal.id;
+      terminal.scope.range.end = firstId;
+    }
+  }
```

Comment:
I stole this from BuildReactiveScopeTerminalsHIR.ts. Should this be a shared utility?

Reply:
yeah!